Web Inspector: Font Panel: CSS font property values marked !important don't get overridden when using the interactive editing controls.

francescorn opened the pull request:
#### 793932ceb15afa7dcbcfa99af428131244f7a15d
<pre>
Web Inspector: Font Panel: CSS font property values marked !important don&apos;t get overridden when using the interactive editing controls.
<a href="https://bugs.webkit.org/show_bug.cgi?id=258975">https://bugs.webkit.org/show_bug.cgi?id=258975</a>

Reviewed by Devin Rousso.

If a CSS class has a !important tag on a property the slider does not work correctly.
This is due to us modifying the inline style, and the inline style is being overriden by
the css class !important property.

We can address this by checking if the !important property is being applied already
during _computeProperty(), and then if it is then append !important in the
writeFontVariation().

Rename resultProperties to resultProperty inside of _computeProperty() because there
is a single property being handled.

Delete _effectivePropertyValueForName function since it is not in use and create it in
_computeProperty instead to be used for resultProperty.isImportant.

* Source/WebInspectorUI/UserInterface/Models/FontStyles.js:
(WI.FontStyles.prototype.writeFontVariation):
(WI.FontStyles):

Canonical link: <a href="https://commits.webkit.org/275389@main">https://commits.webkit.org/275389@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ff503c35b63cc0d9c7f1893626695d60f2c47fbd

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/41390 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/20404 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/43768 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/43955 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/37482 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/23519 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/17734 "Built successfully") | [❌ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/34210 "Found 1 new test failure: media/track/media-element-enqueue-event-crash.html (failure)") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/41964 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/17345 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/35635 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/14871 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/15061 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/36638 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/45298 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/37591 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/36957 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/40702 "Passed tests") | 
| [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/41016 "Build is in progress. Recent messages:") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/16205 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/13285 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/39103 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/17824 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/9350 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/17877 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/17468 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

rcaliman-apple commented on Source/WebInspectorUI/UserInterface/Models/DOMNodeStyles.js:
```diff
@@ -519,7 +519,7 @@ WI.DOMNodeStyles = class DOMNodeStyles extends WI.Object
         var text = payload.text || "";
         var name = payload.name;
-        var value = payload.value || "";
+        var value = (payload.value || "").replace(/\s*!important\s*$/, "");
```
There's no need to modify the value here in the parsing stage when receiving it from the backend. 
This is causing the `inspector/css/modify-inline-style.html` test to fail.

francescorn replied:
Removed it

rcaliman-apple commented on Source/WebInspectorUI/UserInterface/Models/FontStyles.js:
```diff
@@ -138,6 +138,12 @@ WI.FontStyles = class FontStyles
         let cssProperty = this._effectiveWritablePropertyForName(targetPropertyName, createIfMissing);
+
+        if (!targetPropertyValue.includes("!important")) {
+            targetPropertyValue += "!important";
```
This will change *always* add the `!important` suffix to the written property value. 
It should be added only when the property value collected from the page originally was marked with `!important`.

To find identify that, you can update `WI.FontStyles._computeProperty()` to look at the `important` property of the `WI.CSSProperty` collected into `WI.DOMNodeStyles`.

Like this: 
```js
// returns true if the CSS property was authored with `!important` in its value (which exists on WI.CSSProperty._rawValue)
style.domNodeStyle.effectivePropertyForName(name).important
```

Update the object returned by `WI.FontStyles._computeProperty` with an additional property, like `isImportant` or similar.

>[!NOTE]
>The name of that return object is misleadingly pluralised now as `resultProperties`. It should be `resultProperty` since that method operates on a single property. Probably a typo from when it was written.

If you follow that return object, you'll notice it gets collected with the other ones for other CSS properties in `WI.FontStyles._propertiesMap`. 

That's the map you can then check against in `WI.FontStyles.writeFontVariation()` to check whether the value of the CSS property you're trying to update was originally marked important. If so, concatenate the "!important" to `targetPropertyValue`.

You don't have to make any changes outside of `WI.FontStyles` to fix this bug. Please don't change `WI.CSSProperty`

rcaliman-apple commented on Source/WebInspectorUI/UserInterface/Models/FontStyles.js:
```diff
@@ -138,6 +138,12 @@ WI.FontStyles = class FontStyles
+
+        if (!targetPropertyValue.includes("!important")) {
+            targetPropertyValue += "!important";
+            cssProperty.inspectorOverrideImportant = true;
```
`inspectorOverrideImportant` does not exist on `WI.CSSProperty`. Please remove.

rcaliman-apple commented on Source/WebInspectorUI/UserInterface/Models/FontStyles.js:
```diff
@@ -312,6 +316,9 @@ WI.FontStyles = class FontStyles
             }
         }
 
+        // returns true if the CSS property was authored with `!important` in its value (which exists on WI.CSSProperty._rawValue)
```
The code comment is redundant. Please remove it. I had put it in the code snippet in my review as an explanation for you

francescorn replied:
Removed

rcaliman-apple commented on Source/WebInspectorUI/UserInterface/Models/FontStyles.js:
```diff
@@ -312,6 +316,9 @@ WI.FontStyles = class FontStyles
         }
 
+        // returns true if the CSS property was authored with `!important` in its value (which exists on WI.CSSProperty._rawValue)
+        resultProperties.isImportant = style.domNodeStyle.effectivePropertyForName(name)?.important;
```
Please rename `resultProperties` to `resultProperty` inside this method. The plural form is a typo from the original patch. This method deals with a single property.

rcaliman-apple reviewed (commented):
Thank you for fixing this.

dcrousso commented on Source/WebInspectorUI/UserInterface/Models/FontStyles.js:
```diff
@@ -138,6 +138,10 @@ WI.FontStyles = class FontStyles
         let cssProperty = this._effectiveWritablePropertyForName(targetPropertyName, createIfMissing);
+
+        if (this._propertiesMap.get(targetPropertyName)?.isImportant)
+            targetPropertyValue += " !important";
```
NIT: are we 100% sure that `targetPropertyValue` can't already have `!important` in it?

rcaliman-apple replied:
The `value` that ends up here is the value of the `<input type="range">` elements in the Fonts sidebar. 
It gets there from `WI.CSSProperty.value` which is guaranteed to strip `!important`:

https://github.com/WebKit/WebKit/blob/aef2df021f711fd6b5e073d41fac755bb607dbcb/Source/WebInspectorUI/UserInterface/Models/CSSProperty.js#L370-L376

`targetPropertyValue` is either that `value` as string or a transformed string according to this mapping:
https://github.com/WebKit/WebKit/blob/aef2df021f711fd6b5e073d41fac755bb607dbcb/Source/WebInspectorUI/UserInterface/Models/FontStyles.js#L73-L85

There's no `!important` ending up here. That's why we need to append `!important` ourselves when writing back to the stylesheet.

dcrousso commented on Source/WebInspectorUI/UserInterface/Models/FontStyles.js:
```diff
@@ -304,15 +308,17 @@ WI.FontStyles = class FontStyles
         }
 
-        return resultProperties;
+        resultProperty.isImportant = style.domNodeStyle.effectivePropertyForName(name)?.important;
```
FYI we're already calling this inside `_effectivePropertyValueForName`, so i think we could/should avoid that
```js
let effectivePropertyForName = style.domNodeStyle.effectivePropertyForName(name);

let value = effectivePropertyForName?.value;

...

resultProperty.isImportant = effectivePropertyForName?.important ?? false;

```

francescorn replied:
Added

francescorn replied:
We need to keep `return resultProperties` included

francescorn replied:
Removing `return resultProperties` was causing web inspector test cases to fail, so I added it back

francescorn committed:
Web Inspector: Font Panel: CSS font property values marked !important don't get overridden when using the interactive editing controls.
https://bugs.webkit.org/show_bug.cgi?id=258975

Reviewed by Devin Rousso.

If a CSS class has a !important tag on a property the slider does not work correctly.
This is due to us modifying the inline style, and the inline style is being overriden by
the css class !important property.

We can address this by checking if the !important property is being applied already
during _computeProperty(), and then if it is then append !important in the
writeFontVariation().

Rename resultProperties to resultProperty inside of _computeProperty() because there
is a single property being handled.

Delete _effectivePropertyValueForName function since it is not in use and create it in
_computeProperty instead to be used for resultProperty.isImportant.

* Source/WebInspectorUI/UserInterface/Models/FontStyles.js:
(WI.FontStyles.prototype.writeFontVariation):
(WI.FontStyles):

Canonical link: https://commits.webkit.org/275389@main
